skimage/morphology/_flood_fill.py

```diff
@@ -230,22 +230,29 @@ def flood(image, seed_point, *, footprint=None, connectivity=None,
     seed_value = image[seed_point]
     seed_point = tuple(np.asarray(seed_point) % image.shape)
 
-    footprint = _resolve_neighborhood(footprint, connectivity, image.ndim)
+    footprint = _resolve_neighborhood(
+        footprint, connectivity, image.ndim, enforce_adjacency=False)
+    center = np.array(footprint.shape) // 2
+
+    # Compute pad count as the maximum offset to neighbors across all axes.
+    neighbor_indices = np.array(np.nonzero(footprint)).T
+    neighbor_coordinate_offsets = np.ravel(neighbor_indices - center)
+    pad_count = np.max(np.abs(neighbor_coordinate_offsets))
 
     # Must annotate borders
-    working_image = np.pad(image, 1, mode='constant',
+    working_image = np.pad(image, pad_count, mode='constant',
                            constant_values=image.min())
 
     # Stride-aware neighbors - works for both C- and Fortran-contiguity
-    ravelled_seed_idx = np.ravel_multi_index([i + 1 for i in seed_point],
-                                             working_image.shape, order=order)
+    ravelled_seed_idx = np.ravel_multi_index(
+        [i + pad_count for i in seed_point], working_image.shape, order=order)
     neighbor_offsets = _offsets_to_raveled_neighbors(
-        working_image.shape, footprint, center=((1,) * image.ndim),
+        working_image.shape, footprint, center=center,
         order=order)
 
     # Use a set of flags; see _flood_fill_cy.pyx for meanings
-    flags = np.zeros(working_image.shape, dtype=np.uint8, order=order)
-    _set_border_values(flags, value=2)
+    flags = np.zeros(image.shape, dtype=np.uint8, order=order)
+    flags = np.pad(flags, pad_count, constant_values=2)
```

Comment:
I proposed some changes at https://github.com/jagao99/scikit-image/pull/1 that allow keeping use of `_set_border_values` here for efficiency. Otherwise `np.pad` will generate a whole new array and copy all values of flags into it which is a bit of extra memory and performance overhead.

If you merge that PR in your repo, the changes will automatically show up here.

Reply:
Done. Thanks for the neat performance improvements!